test/testpushnotifications.cpp

```diff
@@ -0,0 +1,351 @@
+
+    account->setCapabilities(capabilitiesMap);
+
+    return account;
+}
+
+std::unique_ptr<OCC::PushNotifications> createPushNotifications(OCC::Account *account)
+{
+    auto pushNotifications = std::make_unique<OCC::PushNotifications>(account);
+    pushNotifications->setReconnectTimerInterval(0);
+
+    return std::move(pushNotifications);
+}
+
+class TestPushNotifications : public QObject
+{
+    Q_OBJECT
+
+private slots:
+    void testSetup_correctCredentials_authenticatedOnWebSocket()
+    {
+        const QString user = "user";
+        const QString password = "password";
+        MockWebSocketServer mockServer;
+        QSignalSpy spy(&mockServer, &MockWebSocketServer::processTextMessage);
+        QVERIFY(spy.isValid());
+        auto account = createAccount();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account.data());
+
+        // Setup push notifications
+        pushNotifications->setup();
+
+        // Wait for authentication
+        spy.wait();
+
+        // Right authentication data should be sent
+        QCOMPARE(spy.count(), 2);
+
+        const auto userSent = spy.at(0).at(1).toString();
+        const auto passwordSent = spy.at(1).at(1).toString();
+
+        QCOMPARE(userSent, user);
+        QCOMPARE(passwordSent, password);
+    }
+
+    void testOnWebSocketTextMessageReceived_notifyFileMessage_emitFilesChanged()
+    {
+        const QString user = "user";
+        const QString password = "password";
+        MockWebSocketServer mockServer;
+        QSignalSpy processTextMessageSpy(&mockServer, &MockWebSocketServer::processTextMessage);
+        QVERIFY(processTextMessageSpy.isValid());
+
+        auto account = createAccount().data();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account);
+        QSignalSpy filesChangedSpy(pushNotifications.get(), &OCC::PushNotifications::filesChanged);
+        QVERIFY(filesChangedSpy.isValid());
+
+        // Setup push notifications
+        pushNotifications->setup();
+
+        // Wait for authentication and then send notify_file push notification
+        processTextMessageSpy.wait();
+        QCOMPARE(processTextMessageSpy.count(), 2);
+        const auto socket = processTextMessageSpy.at(0).at(0).value<QWebSocket *>();
+        socket->sendTextMessage("notify_file");
+
+        // filesChanged signal should be emitted
+        QVERIFY(filesChangedSpy.wait());
+        QCOMPARE(filesChangedSpy.count(), 1);
+        // FIXME: This fails because accountFilesChanged is always nullptr. What do I wrong?
+        // I have checked that in a normal signal handler I get the correct account pointer.
+        // auto accountFilesChanged = filesChangedSpy.at(0).at(0).value<OCC::Account *>();
```

Comment:
Is OCC::Account * properly registered to the meta type system? What if you try to get a QObject * out instead? You're sure you're not getting a warning at runtime around that type being used in a QVariant?

I suspect a `qRegisterMetaType()` call missing in that test suite. Indeed AccountPtr and AccountState* are registered but not Account*.

Alternatively maybe use AccountPtr all the way in PushNotifications? Doesn't have my favor design wise but this seems to be pretty much the common case everywhere else in libsync...

Reply:
It was not correctly registered `qRegisterMetaType()` with `Q_DECLARE_METATYPE` did the trick.